### ISSUE TITLE
Nondeterminism in FastHierarchy with phantom classes

### ISSUE THREAD

**Issue by jtoman** [Sable/soot#997]
The FastHierarchy uses intervals to efficiently represent subclassing relationships. When the FH is created, these intervals are computed by walking type hierarchy from Object, where the object interval begins at 0. Then, for each phantom class, FH walks those classes' type hierarchies to compute the intervals, again beginning at 0. Depending on the order classes are iterated, certain phantom classes can be assigned intervals that just happen to be subranges of intervals assigned to non-phantom classes. Across differnet runs, FH can nondeterministically declare two classes subtypes of one another.

I'm not sure how to fix this bug. The easy solution would be to start each phantom classes' interval from the upper bound of the previous hierarchy's interval, but this is technically unsound as phantom classes may very well be subtypes of non-phantom classes.

**Comment by mbenz89** [comment]
Hi John. 

I got a very packed week and will be on vacation afterward, so please indulge me if this may take a few weeks to get resolved.

**Comment by jtoman** [comment]
Yeah no problem. FWIW, I worked around this by ensuring all classes were in hierarchy, and as I mentioned in person, for large enough sets of classes the chances that any given subtype check will hit this bug are relatively small.

**Comment by jpstotz** [comment]
I also ran into this problem. To make it works I encountered the case where `canStoreClass` is called with two classes that are both phantom (real phantoms without a hierarchy -> no super class set). In such a case the result is always `true`!

Shouldn't we add a check in the FastHierarchy for cases where one or both arguments are "phantom" classes without hierarchy/super class?
